File: Design-Patterns-Using-Cpp/Builder/BuilderPattern.cpp
#include <iostream>
#include <sstream>

// Define the Product class with two parts
class Product {
private:
    std::string partA;
    std::string partB;

public:
    Product(const std::string& A, const std::string& B) : partA(A), partB(B) {}

    void setPartA(const std::string& A){
        this->partA = A;
    }

    void setPartB(const std::string& B){
        this->partB = B;
    }

    friend std::ostream& operator<<(std::ostream& os, const Product& product) {
        os << "Product: (" << product.partA << ", " << product.partB << ")";
        return os;
    }
};

// Define an abstract class called Builder
class Builder {
protected:
    Product product;

public:
    Builder() : product("A default", "B default") {}

    virtual Builder* setPartA(const std::string& A) = 0;
    virtual Builder* setPartB(const std::string& B) = 0;

    Product getProduct() {
        Product temp = this->product;
        this->product = Product("A default", "B default"); // assign new product.
        return temp;
    }
};

// Define a ConcreteBuilder class that extends Builder
class ConcreteBuilder : public Builder {
public:
    Builder* setPartA(const std::string& A) override {
        this->product.setPartA(A);
        return this;
    }

    Builder* setPartB(const std::string& B) override {
        this->product.setPartB(B);
        return this;
    }
};

// Define a Director class that takes a builder object as a parameter
class Director {
private:
    Builder* builder;

public:
    Director(Builder* builder) : builder(builder) {}

    Product construct() {
        return this->builder->setPartA("A1")->setPartB("B1")->getProduct();
    }

    Product construct2() {
        this->builder->setPartA("A2");
        this->builder->setPartB("B2");
        return this->builder->getProduct();
    }

    Product construct3() {
        return this->builder->setPartA("A3")->getProduct();
    }
};

// Client code
int main() {
    Builder* builder = new ConcreteBuilder();
    Director director(builder);

    Product product = director.construct();
    std::cout << product << std::endl;

    Product product2 = director.construct2();
    std::cout << product2 << std::endl;

    Product product3 = director.construct3();
    std::cout << product3 << std::endl;

    // Clean up memory
    delete builder;

    return 0;
}

/*
Product: (A1, B1)
Product: (A2, B2)
Product: (A3, B default)
*/

File: Design-Patterns-Using-Kotlin/bin/BuilderPattern.kt
// Define the Product class with two parts
open class Product(private var partA: String, private var partB: String) {
    fun setPartA(A: String) {
        this.partA = A
    }

    fun setPartB(B: String) {
        this.partB = B
    }

    override fun toString(): String {
        return String.format("Product : (%s, %s)", partA, partB)
    }
}

// Define an abstract class called Builder
abstract class Builder {
    protected var prod: Product = Product("A default", "B default")

    abstract fun setPartA(A: String): Builder

    abstract fun setPartB(B: String): Builder

    open fun getProduct(): Product {
        val temp = this.prod
        this.prod = Product("A default", "B default") // assign new product.
        return temp
    }
}

// Define a ConcreteBuilder class that extends Builder
class ConcreteBuilder : Builder() {
    override fun setPartA(A: String): Builder {
        prod.setPartA(A)
        return this
    }

    override fun setPartB(B: String): Builder {
        prod.setPartB(B)
        return this
    }
}

// Define a Director class that takes a builder object as a parameter
class Director(private val builder: Builder) {
    fun construct(): Product {
        return builder.setPartA("A1").setPartB("B1").getProduct()
    }

    fun construct2(): Product {
        builder.setPartA("A2")
        builder.setPartB("B2")
        return builder.getProduct()
    }

    fun construct3(): Product {
        return builder.setPartA("A3").getProduct()
    }
}

// Client code
fun main() {
    val builder: Builder = ConcreteBuilder()
    val director = Director(builder)

    val product = director.construct()
    println(product)

    val product2 = director.construct2()
    println(product2)

    val product3 = director.construct3()
    println(product3)
}
 
/*
Product : (A1, B1)
Product : (A2, B2)
Product : (A3, B default)
 */

